src/qibolab/instruments/zhinst/sweep.py

```diff
@@ -0,0 +1,139 @@
+                            - qubit.readout.local_oscillator.frequency
+                        )
+                    elif ptype is PulseType.DRIVE:
+                        ch = qubit.drive.name
+                        intermediate_frequency = (
+                            qubit.drive_frequency
+                            - qubit.drive.local_oscillator.frequency
+                        )
+                    else:
+                        raise ValueError(
+                            f"Cannot sweep frequency of pulse of type {ptype}, because it does not have associated frequency"
+                        )
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values + intermediate_frequency
+                    )
+                    channel_sweeps.append((ch, sweeper.parameter, sweep_param))
+                elif (
+                    pulse.type is PulseType.READOUT
+                    and sweeper.parameter is Parameter.amplitude
+                ):
+                    sweep_param = lo.SweepParameter(
+                        values=sweeper.values / max(sweeper.values)
+                    )
+                    channel_sweeps.append(
+                        (
+                            measure_channel_name(qubits[pulse.qubit]),
+                            sweeper.parameter,
+                            sweep_param,
+                        )
+                    )
+                else:
+                    sweep_param = lo.SweepParameter(values=copy(sweeper.values))
+                    pulse_sweeps.append((pulse, sweeper.parameter, sweep_param))
+                parallel_sweeps.append((sweeper, sweep_param))
+
+            for qubit in sweeper.qubits or []:
+                if sweeper.parameter is not Parameter.bias:
+                    raise ValueError(
+                        f"Sweeping {sweeper.parameter.name} for {qubit} is not supported"
+                    )
+                sweep_param = lo.SweepParameter(
+                    values=sweeper.values + qubit.flux.offset
+                )
+                channel_sweeps.append((qubit.flux.name, sweeper.parameter, sweep_param))
+                parallel_sweeps.append((sweeper, sweep_param))
+
+            for coupler in sweeper.couplers or []:
+                if sweeper.parameter is not Parameter.bias:
+                    raise ValueError(
+                        f"Sweeping {sweeper.parameter.name} for {coupler} is not supported"
+                    )
+                sweep_param = lo.SweepParameter(
+                    values=sweeper.values + coupler.flux.offset
+                )
+                channel_sweeps.append(
+                    (coupler.flux.name, sweeper.parameter, sweep_param)
+                )
+                parallel_sweeps.append((sweeper, sweep_param))
+
+        self._pulse_sweeps = pulse_sweeps
+        self._channel_sweeps = channel_sweeps
+        self._parallel_sweeps = parallel_sweeps
```

Comment:
Since the attributes are all computed after the parameters, I would suggest the following:
- define the class with `@dataclass`, and use as attributes the current `__init__` parameters (they are not heavy, no need to drop them)
- define current attributes as `@cached_properties` (so you don't have to recompute them, but you can split the code)
- if there are shared computations, split them as separate methods or further internal `cached_properties`

I'm not yet sure all the attributes are actually strictly needed (they seem more like multiple indices, with the same info, but you could compute them lazily...).